src/platform/intel/cavs/lib/clk.c

```diff
@@ -145,24 +129,67 @@ static inline void set_cpu_current_freq_idx(int freq_idx, bool release_unused)
 {
 	select_cpu_clock(freq_idx, true);
 }
+#endif /* !(CONFIG_CAVS_LPRO_ONLY) */
 
 static void platform_clock_low_power_mode(int clock, bool enable)
 {
+#if CONFIG_CAVS_USE_LPRO_IN_WAITI
+	int current_freq_idx = get_current_freq_idx(clock);
+	int freq_idx = *cache_to_uncache(&active_freq_idx);
+
+	if (enable && current_freq_idx != CPU_LPRO_FREQ_IDX)
+		/* LPRO requests are fast, but requests for other ROs
+		 * can take a lot of time. That's why it's better to
+		 * not release active clock just for waiti,
+		 * so they can be switched without delay on wake up.
+		 */
+		select_cpu_clock(CPU_LPRO_FREQ_IDX, false);
+	else if (!enable && current_freq_idx != freq_idx)
+		select_cpu_clock(freq_idx, false);
+#endif /* CONFIG_CAVS_USE_LPRO_IN_WAITI */
 }
-#endif
 
-void platform_clock_on_waiti(void)
+static int clock_platform_set_cpu_freq(int clock, int freq_idx)
 {
+	set_cpu_current_freq_idx(freq_idx, true);
+	return 0;
 }
 
-void platform_clock_on_wakeup(void)
+void platform_clock_on_waiti(void)
 {
+/* do nothing for LPRO_ONLY mode */
+#if !(CONFIG_CAVS_LPRO_ONLY)
+	/* perform the possible dynamic clock switching */
+	if (*cache_to_uncache(&active_freq_idx) != CPU_LPRO_FREQ_IDX &&
+	    !pm_runtime_is_active(PM_RUNTIME_DSP, PLATFORM_MASTER_CORE_ID))
+		set_cpu_current_freq_idx(CPU_LPRO_FREQ_IDX, true);
+	else if (*cache_to_uncache(&active_freq_idx) != CPU_HPRO_FREQ_IDX &&
+		 pm_runtime_is_active(PM_RUNTIME_DSP, PLATFORM_MASTER_CORE_ID))
+		set_cpu_current_freq_idx(CPU_HPRO_FREQ_IDX, true);
+
+#if CONFIG_CAVS_USE_LPRO_IN_WAITI
+	/* check if waiti HPRO->LPRO switching is needed */
+	pm_runtime_put(CORE_HP_CLK, cpu_get_id());
```

Comment:
This version of on_waiti/on_wakeup is removed by the previous commit and restored by this one. Should those two ones be squashed to avoid intermediate commit that has some regression?

Reply:
good point, let me squash them.